### PR TITLE
[RDY] Evaluation in context with dynsem generated interpreter

### PR DESCRIPTION
**Pull request by Balletie**
Relies on the following PRs:
- spoofax-shell/esv#1 for specifying shell configuration parameters in ESV.
- spoofax-shell/spoofax#1 for the ShellFacet that is parsed from the ESV.
- spoofax-shell/simpl#1 for the shell configuration of the simpl language.
- metaborg/dynsem#55 for the refactored DynSem entrypoint.

### CODE REVIEW

File: org.metaborg.spoofax.shell.core/ruleset.xml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    xmlns="http://pmd.sourceforge.net/ruleset/2.0.0"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://pmd.sourceforge.net/ruleset/2.0.0 http://pmd.sourceforge.net/ruleset_2_0_0.xsd">
+  <description>
+  Overriding some rule properties.
+  </description>
+
+  <rule ref="rulesets/java/empty.xml/EmptyCatchBlock">
+    <properties>
+      <property name="allowCommentedBlocks" value="true"/>
```
Comment (Hjdskes) [inline org.metaborg.spoofax.shell.core/ruleset.xml]: Why is this needed?

File: org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/commands/EvaluateCommand.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+     * Analyze before invoking the evaluation strategy.
+     */
+    private class AnalyzedInvocationStrategy implements EvaluationInvocationStrategy {
+        @Override
+        public EvaluateResult performEvaluation(IContext context, ParseResult parsed,
+                                                IEvaluationStrategy evalStrategy)
+            throws MetaborgException {
+            AnalyzeResult analyzed = analyzeCommand.analyze(parsed);
+            IStrategoTerm ast = evalStrategy.evaluate(analyzed, context);
+            // TODO: Normally this is done in the unit service, but since this is my own Unit this
```
Comment (Hjdskes) [inline org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/commands/EvaluateCommand.java]: So this would require a change in Spoofax itself? If so, is this doable for us, or does it work good enough as-is / does Spoofax itself have no use for it besides the shell?

File: org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/commands/EvaluateCommand.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+            // is not possible.
+            EvaluateUnit<ISpoofaxAnalyzeUnit> unit =
+                new EvaluateUnit<ISpoofaxAnalyzeUnit>(((UnitWrapper) analyzed.unit()).unit, ast,
```
Comment (Hjdskes) [inline org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/commands/EvaluateCommand.java]: This cast (and the one below) are known to always work?

File: org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/core/ClassPathInterpreterLoader.java [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+        }
+    }
+
+    /* Loads the required configurations from the dynsem.properties file parsed as a Properties
+     * object. */
+    private void loadDynSemProperties(ILanguageImpl langImpl) throws InterpreterLoadException {
+        FileObject dynSemPropertiesFile = findDynSemPropertiesFileForLanguage(langImpl);
+
+        Properties dynSemProperties = new Properties();
+        InputStream in;
```
Comment (Hjdskes) [inline org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/core/ClassPathInterpreterLoader.java]: 1. This can go inside the `try` (it's not used after the `try/catch` block).
2. Shouldn't this `InputStream` be closed?

File: org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/core/DynSemEvaluationStrategy.java [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    }
+
+    private @Nullable Value lookupConRule(IStrategoTerm input) {
+        IStrategoConstructor inputCtor = ((IStrategoAppl) input).getConstructor();
+        String ctorName = inputCtor.getName();
+        int arity = inputCtor.getArity();
+
+        Value rule =
+            polyglotEngine.findGlobalSymbol(RuleRegistry.makeKey("shell", ctorName, arity));
+        return rule;
```
Comment (Hjdskes) [inline org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/core/DynSemEvaluationStrategy.java:141]: You can use `return polyglotEngine.findGlobalSymbol(...);` here.

File: org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/core/IInterpreterLoader.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+     * @return The DynSem entrypoint for the interpreter.
+     * @throws InterpreterLoadException
+     *             When loading results in an error.
+     */
+    PolyglotEngine loadInterpreterForLanguage(ILanguageImpl langImpl)
+        throws InterpreterLoadException;
+
+    /**
+     * Returns a {@link ITerm} corresponding to the given {@link IStrategoTerm}. Relies on
+     * {@link StrategoUtil#getSortForTerm(IStrategoTerm)} for getting the term's sort.
```
Comment (Hjdskes) [inline org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/core/IInterpreterLoader.java:31]: Isn't this an implementation detail?

File: org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/core/IInterpreterLoader.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+     *            The {@link IStrategoTerm}.
+     * @return The corresponding {@link ITerm}.
+     * @throws InterpreterLoadException
+     *             When finding the {@link ITerm} results in failure.
+     */
+    ITerm getProgramTerm(IStrategoTerm term) throws InterpreterLoadException;
+
+    /**
+     * Returns the transformer which optionally transforms a given {@link IStrategoTerm} before
+     * evaluation. By default this is the {@link ITermTransformer.IDENTITY identity transformation}.
```
Comment (Hjdskes) [inline org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/core/IInterpreterLoader.java:43]: Same as above; isn't this an implementation detail?

File: org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/output/EvaluateResult.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    @Override
+    public List<IMessage> messages() {
+        return Collections.emptyList();
+    }
+
+    @Override
+    public StyledText styled() {
+        return toString(unit().ast());
```
Comment (Hjdskes) [inline org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/output/EvaluateResult.java]: Is this result always valid? I guess so considering `valid()`, but just making sure.

File: org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/output/EvaluateResult.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    @Override
+    public Optional<IContext> context() {
+        return Optional.of(unit().context());
+    }
+
+    @Override
+    public List<IMessage> messages() {
+        return Collections.emptyList();
```
Comment (Hjdskes) [inline org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/output/EvaluateResult.java]: There is never a message to display? I guess not since it's apparently always valid, but I'm just checking.

File: org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/util/StrategoUtil.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+import org.spoofax.interpreter.terms.IStrategoTerm;
+import org.spoofax.jsglr.client.imploder.ImploderAttachment;
+
+/**
+ * Contains utility functions for dealing with {@link IStrategoTerm}s.
+ */
+public final class StrategoUtil {
+
+    /* private due to checkstyle. */
```
Comment (Hjdskes) [inline org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/util/StrategoUtil.java:17]: I don't think this comment is really necessary :P

File: org.metaborg.spoofax.shell.core/src/test/java/org/metaborg/spoofax/shell/core/ClassPathInterpreterLoaderTest.java [hunk 11]
```diff
@@ -0,0 +1,341 @@
...
+        this.expectedExceptionMessage = expectedExceptionMessage;
+        this.expectedExceptionCauseClass = expectedExceptionCauseClass;
+    }
+
+    /**
+     * Set up mocks.
+     */
+    @Before
+    public void setUp() {
+        mockLangImpl = Mockito.mock(ILanguageImpl.class);
```
Comment (Unia) [inline org.metaborg.spoofax.shell.core/src/test/java/org/metaborg/spoofax/shell/core/ClassPathInterpreterLoaderTest.java]: You can use the `@Mock` annotation instead of doing this.

File: org.metaborg.spoofax.shell.core/src/test/java/org/metaborg/spoofax/shell/core/ClassPathInterpreterLoaderTest.java [hunk 12]
```diff
@@ -0,0 +1,341 @@
...
+                printWriter.println("project.javapackage = org.metaborg.spoofax.shell.core");
+                printWriter
+                    .println("source.langname = ClassPathInterpreterLoaderTest$TestCPLoader");
+                printWriter.flush();
+                printWriter.close();
+                dynSemProperties.close();
+                List<FileObject> res = new ArrayList<FileObject>(1);
+                res.add(rootDir);
+                return res;
+            } catch (FileSystemException e) {
```
Comment (Hjdskes) [inline org.metaborg.spoofax.shell.core/src/test/java/org/metaborg/spoofax/shell/core/ClassPathInterpreterLoaderTest.java:157]: You can just throw these exceptions; when caught by JUnit they will automatically make the test fail.

File: org.metaborg.spoofax.shell.core/src/test/java/org/metaborg/spoofax/shell/core/ClassPathInterpreterLoaderTest.java [hunk 13]
```diff
@@ -0,0 +1,341 @@
...
+                PrintWriter printWriter =
+                    new PrintWriter(noDynSemProperties.getContent().getOutputStream());
+                printWriter.println("blabla");
+                printWriter.flush();
+                printWriter.close();
+                noDynSemProperties.close();
+                List<FileObject> res = new ArrayList<FileObject>(1);
+                res.add(rootDir);
+                return res;
+            } catch (FileSystemException e) {
```
Comment (Hjdskes) [inline org.metaborg.spoofax.shell.core/src/test/java/org/metaborg/spoofax/shell/core/ClassPathInterpreterLoaderTest.java:180]: Same as above.

File: org.metaborg.spoofax.shell.core/src/test/java/org/metaborg/spoofax/shell/core/ClassPathInterpreterLoaderTest.java [hunk 14]
```diff
@@ -0,0 +1,341 @@
...
+                someRootDir.createFolder();
+                FileObject anotherRootDir = rootDir.resolveFile("anotherRootDir/");
+                anotherRootDir.createFolder();
+                List<FileObject> missingDynSemProperties =
+                    missingDynSemProperties(anotherRootDir).get();
+                List<FileObject> goodWeatherLocations = goodWeatherLocations(someRootDir).get();
+                missingDynSemProperties.addAll(goodWeatherLocations);
+
+                return missingDynSemProperties;
+            } catch (FileSystemException e) {
```
Comment (Hjdskes) [inline org.metaborg.spoofax.shell.core/src/test/java/org/metaborg/spoofax/shell/core/ClassPathInterpreterLoaderTest.java:202]: Same as above.

File: org.metaborg.spoofax.shell.core/src/test/java/org/metaborg/spoofax/shell/core/ClassPathInterpreterLoaderTest.java [hunk 15]
```diff
@@ -0,0 +1,341 @@
...
+                    new PrintWriter(dynSemProperties.getContent().getOutputStream());
+                printWriter.println("project.javapackage = non.existing.package");
+                printWriter.println("source.langname = NonExistingLanguage");
+                printWriter.flush();
+                printWriter.close();
+                dynSemProperties.close();
+                List<FileObject> res = new ArrayList<FileObject>(1);
+                res.add(rootDir);
+                return res;
+            } catch (FileSystemException e) {
```
Comment (Hjdskes) [inline org.metaborg.spoofax.shell.core/src/test/java/org/metaborg/spoofax/shell/core/ClassPathInterpreterLoaderTest.java:225]: Same as above.

File: org.metaborg.spoofax.shell.core/src/test/java/org/metaborg/spoofax/shell/core/ClassPathInterpreterLoaderTest.java [hunk 16]
```diff
@@ -0,0 +1,341 @@
...
+            // Test finding a rule.
+            Value testRuleValue =
+                engine.findGlobalSymbol(RuleRegistry.makeKey("testrule", "TestCtor", 0));
+            DynSemRule testRule = testRuleValue.as(DynSemRule.class);
+
+            assertEquals(testRule.getRuleTarget(), MOCK_RULE_ROOT);
+
+            // Assert that the parse method of the test language class was reached.
+            assertTrue(reachedLanguageParse);
+        } catch (IOException e) {
```
Comment (Hjdskes) [inline org.metaborg.spoofax.shell.core/src/test/java/org/metaborg/spoofax/shell/core/ClassPathInterpreterLoaderTest.java:260]: Same as above.

File: org.metaborg.spoofax.shell.core/src/test/java/org/metaborg/spoofax/shell/core/DynSemEvaluationStrategyTest.java [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+            // First invocation causes an update in the environment.
+            IStrategoTerm firstResult = evalStrategy
+                .evaluate(analyzeResultMock, mock(IContext.class, Mockito.RETURNS_MOCKS));
+            assertTrue(firstResult.toString().contains("0"));
+
+            // Second invocation has a different result due to a different environment.
+            IStrategoTerm secondResult = evalStrategy
+                .evaluate(analyzeResultMock, mock(IContext.class, Mockito.RETURNS_MOCKS));
+            assertTrue(secondResult.toString().contains(String.valueOf(DEFAULT_SHELL_RULE_ANSWER)));
+        } catch (MetaborgException e) {
```
Comment (Hjdskes) [inline org.metaborg.spoofax.shell.core/src/test/java/org/metaborg/spoofax/shell/core/DynSemEvaluationStrategyTest.java:126]: Just throw 'em (aka same as above).

**Comment by Hjdskes** [comment]
Really nice work @Balletie! Is it necessary for me / @gfokkema to review the PRs you linked to, as well? I glanced over them quickly and they look good to me.

Before we merge this, I'd like to see the build pass, too.

File: org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/commands/EvaluateCommand.java (hunk 2 above)
Comment (Balletie) [inline org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/commands/EvaluateCommand.java]: I'm not sure if an `EvaluateUnit` is something they want, so that's why I did not implement this in the fork.

File: org.metaborg.spoofax.shell.core/ruleset.xml (hunk 1 above)
Comment (Balletie) [inline org.metaborg.spoofax.shell.core/ruleset.xml]: For this line: https://github.com/spoofax-shell/spoofax-shell/pull/20/files#diff-94259d4a4971574c51859b462d3317a7R139

File: org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/commands/EvaluateCommand.java (hunk 3 above)
Comment (Balletie) [inline org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/commands/EvaluateCommand.java]: They cast it in this way in the `UnitService` as well.

File: org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/core/IInterpreterLoader.java (hunk 7 above)
Comment (Balletie) [inline org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/core/IInterpreterLoader.java:43]: This is just the way it is generated by DynSem, I don't set this default myself.

File: org.metaborg.spoofax.shell.core/src/test/java/org/metaborg/spoofax/shell/core/DynSemEvaluationStrategyTest.java (hunk 17 above)
Comment (Balletie) [inline org.metaborg.spoofax.shell.core/src/test/java/org/metaborg/spoofax/shell/core/DynSemEvaluationStrategyTest.java:126]: I'm making an assertion on the exception message, so in this case I want to catch the exception.

File: org.metaborg.spoofax.shell.core/src/test/java/org/metaborg/spoofax/shell/core/ClassPathInterpreterLoaderTest.java (hunk 12 above)
Comment (Balletie) [inline org.metaborg.spoofax.shell.core/src/test/java/org/metaborg/spoofax/shell/core/ClassPathInterpreterLoaderTest.java:157]: Cannot throw in a lambda.
